compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -0,0 +1,306 @@
+use crate::{ImplTraitPosition, ResolverAstLoweringExt};
+
+use super::{ImplTraitContext, LoweringContext, ParamMode};
+
+use ast::visit::Visitor;
+use hir::def::{DefKind, PartialRes, Res};
+use hir::{BodyId, HirId};
+use rustc_ast as ast;
+use rustc_ast::*;
+use rustc_hir as hir;
+use rustc_hir::def_id::DefId;
+use rustc_middle::span_bug;
+use rustc_middle::ty::ResolverAstLowering;
+use rustc_span::symbol::kw;
+use rustc_span::{symbol::Ident, Span};
+use rustc_target::spec::abi;
+use std::iter;
+
+pub struct DelegationResults<'hir> {
+    pub body_id: hir::BodyId,
+    pub sig: hir::FnSig<'hir>,
+    pub generics: &'hir hir::Generics<'hir>,
+}
+
+impl<'hir> LoweringContext<'_, 'hir> {
+    pub fn delegation_has_self(&self, node_id: NodeId, span: Span) -> bool {
+        let res_id = self.get_delegation_res_id(node_id, span);
+        let Some(res_id) = res_id else {
+            return false;
+        };
+
+        if let Some(local_res_id) = res_id.as_local() {
+            self.resolver.has_self.contains(&local_res_id)
+        } else {
+            match self.tcx.def_kind(res_id) {
+                DefKind::Fn => false,
+                DefKind::AssocFn => self.tcx.associated_item(res_id).fn_has_self_parameter,
+                _ => span_bug!(span, "unexpected DefKind for delegation item"),
+            }
+        }
+    }
+
+    pub fn lower_delegation(
+        &mut self,
+        delegation: &Delegation,
+        node_id: NodeId,
+        span: Span,
+    ) -> DelegationResults<'hir> {
+        let res_id = self.get_delegation_res_id(node_id, span);
+        let Some(res_id) = res_id else {
+            return self.generate_delegation_error(span);
+        };
+
+        let decl = self.lower_delegation_decl(res_id, span);
+        let sig = self.lower_delegation_sig(span, decl);
+        let body_id = self.lower_delegation_body(sig.decl, delegation, node_id, span);
+
+        let generics = self.lower_delegation_generics(span);
+        DelegationResults { body_id, sig, generics }
+    }
+
+    fn get_delegation_res_id(&self, node_id: NodeId, span: Span) -> Option<DefId> {
+        let res = self
+            .resolver
+            .get_partial_res(node_id)
+            .map(|or| or.full_res().map(|r| r.opt_def_id()).unwrap_or(None))
+            .unwrap_or(None);
```

Comment:
Resolver always sets resolutions for these ids, even in case of errors, no?
And they are never partial due to `defer_to_typeck == false` for `PathSource::Delegation`.
We should be able to get rid of all these `Option`s.

```rust
let res = self.resolver.get_partial_res(node_id).unwrap().expect_full_res().opt_def_id();
```

Reply:
No. `check_trait_item` will not set the item id resolution in case of errors.